subsys/net/ip/tcp2_priv.h

```diff
@@ -35,26 +37,55 @@
 #if IS_ENABLED(CONFIG_NET_TEST_PROTOCOL)
 #define tcp_pkt_alloc(_conn, _len)					\
 ({									\
-	sa_family_t _family = net_context_get_family((_conn)->context);	\
-	struct net_pkt *_pkt = net_pkt_alloc_with_buffer((_conn)->iface,\
-							 (_len),	\
-							 _family,	\
-							 IPPROTO_TCP,	\
-							 K_NO_WAIT);	\
+	struct net_pkt *_pkt;						\
+									\
+	if ((_len) > 0) {						\
+		_pkt = net_pkt_alloc_with_buffer(			\
+			(_conn)->iface,					\
+			(_len),						\
+			net_context_get_family((_conn)->context),	\
+			IPPROTO_TCP,					\
+			TCP_PKT_ALLOC_TIMEOUT);				\
+	} else {							\
+		if (!(_conn)->unacked_data) {				\
+			_pkt = net_pkt_alloc(TCP_PKT_ALLOC_TIMEOUT);	\
+		} else {						\
+			_pkt = (_conn)->unacked_data;			\
+			goto skip_alloc;				\
+		}							\
+	}								\
 									\
 	tp_pkt_alloc(_pkt, tp_basename(__FILE__), __LINE__);		\
 									\
+skip_alloc:								\
 	_pkt;								\
 })
 #define tcp_pkt_clone(_pkt) tp_pkt_clone(_pkt, tp_basename(__FILE__), __LINE__)
 #define tcp_pkt_unref(_pkt) tp_pkt_unref(_pkt, tp_basename(__FILE__), __LINE__)
 #else
 #define tcp_pkt_alloc(_conn, _len)					\
-	net_pkt_alloc_with_buffer((_conn)->iface, (_len),		\
-				  net_context_get_family((_conn)->context), \
-				  IPPROTO_TCP, K_NO_WAIT)
+({									\
+	struct net_pkt *_pkt;						\
+									\
+	if ((_len) > 0) {						\
+		_pkt = net_pkt_alloc_with_buffer(			\
+			(_conn)->iface,					\
+			(_len),						\
+			net_context_get_family((_conn)->context),	\
+			IPPROTO_TCP,					\
+			TCP_PKT_ALLOC_TIMEOUT);				\
+	} else {							\
+		if (!(_conn)->unacked_data) {				\
```

Comment:
Same here: Shall we just alloc empty packet in tcp_conn_alloc() and drop working with (_conn)->unacked_data here?